quickwit/quickwit-search/src/leaf.rs

```diff
@@ -394,15 +399,283 @@ async fn leaf_search_single_split(
 /// or applying date range when the range covers the entire split.
-fn rewrite_request(search_request: &mut SearchRequest, split: &SplitIdAndFooterOffsets) {
+fn rewrite_request(
+    search_request: &mut SearchRequest,
+    split: &SplitIdAndFooterOffsets,
+    timestamp_field: Option<&str>,
+) {
     if search_request.max_hits == 0 {
         search_request.sort_fields = Vec::new();
     }
-    rewrite_start_end_time_bounds(
-        &mut search_request.start_timestamp,
-        &mut search_request.end_timestamp,
-        split,
-    )
+    if let Some(timestamp_field) = timestamp_field {
+        remove_redundant_timestamp_range(search_request, split, timestamp_field);
+    }
+}
+
+// equivalent to Bound::map, which is unstable
+pub fn map_bound<T, U>(bound: Bound<T>, f: impl FnOnce(T) -> U) -> Bound<U> {
+    use Bound::*;
+    match bound {
+        Unbounded => Unbounded,
+        Included(x) => Included(f(x)),
+        Excluded(x) => Excluded(f(x)),
+    }
+}
+
+// returns the max of left and right, that isn't unbounded. Useful for making
+// the intersection of lower bound of ranges
+fn max_bound<T: Ord + Copy>(left: Bound<T>, right: Bound<T>) -> Bound<T> {
+    use Bound::*;
+    match (left, right) {
+        (Unbounded, right) => right,
+        (left, Unbounded) => left,
+        (Included(left), Included(right)) => Included(left.max(right)),
+        (Excluded(left), Excluded(right)) => Excluded(left.max(right)),
+        (excluded_total @ Excluded(excluded), included_total @ Included(included)) => {
+            if included > excluded {
+                included_total
+            } else {
+                excluded_total
+            }
+        }
+        (included_total @ Included(included), excluded_total @ Excluded(excluded)) => {
+            if included > excluded {
+                included_total
+            } else {
+                excluded_total
+            }
+        }
+    }
+}
+
+// returns the min of left and right, that isn't unbounded. Useful for making
+// the intersection of upper bound of ranges
+fn min_bound<T: Ord + Copy>(left: Bound<T>, right: Bound<T>) -> Bound<T> {
+    use Bound::*;
+    match (left, right) {
+        (Unbounded, right) => right,
+        (left, Unbounded) => left,
+        (Included(left), Included(right)) => Included(left.min(right)),
+        (Excluded(left), Excluded(right)) => Excluded(left.min(right)),
+        (excluded_total @ Excluded(excluded), included_total @ Included(included)) => {
+            if included < excluded {
+                included_total
+            } else {
+                excluded_total
+            }
+        }
+        (included_total @ Included(included), excluded_total @ Excluded(excluded)) => {
+            if included < excluded {
+                included_total
+            } else {
+                excluded_total
+            }
+        }
+    }
+}
+
+/// remove timestamp range that would be present both in QueryAst and SearchRequest
```

Comment:
you don't tell if ends up expressed in the AST or in the search request in that case